Lambda impl methods static and more stably named

retronym opened the pull request:
The body of lambdas is compiled into a synthetic method
in the enclosing class. Previously, this method was a public
virtual method named `fully$qualified$Class$$anonfun$n`.
For lambdas that didn't capture a `this` reference, a static
method was used.

This commit changes two aspects.

Firstly, all lambda impl methods are now emitted static.
An extra parameter is added to those that require a this
reference.

This is an improvement as it:
- allows, shorter, more readable names for the lambda impl method
- avoids pollution of the vtable of the class. Note that javac uses
  private instance methods, rather than public static methods. If
  we followed its lead, we would be unable to support important use
  cases in our inliner

Secondly, the name of the enclosing method has been included in
the name of the lambda impl method to improve debuggability and
to improve serialization compatibility. The serialization improvement
comes from the way that fresh names for the impl methods are
allocated: adding or removing lambdas in methods not named "foo" won't
change the numbering of the `anonfun$foo$n` impl methods from methods
named "foo". This is in line with user expectations about anonymous
class and lambda serialization stability. Brian Goetz has described
this tricky area well in:

  http://cr.openjdk.java.net/~briangoetz/eg-attachments/lambda-serialization.html

This commit doesn't go as far a Javac, we don't use the hash of the
lambda type info, param names, etc to map to a lambda impl method name.
As such, we are more prone to the type-1 and -2 failures described there.
However, our Scala 2.11.8 has similar characteristics, so we aren't going
backwards.

Special case in the naming: Use `"new"` rather than `"<init>"` for constructor enclosed
lambdas, as javac does.

Review by @lrytz

I am still looking at one of your earlier review comments:
- [x] don't we need a recursive call to nonLocalEnclosingMember here? i.e., `if (sym.isLocalToBlock) nonLocalEnclosingMember(sym.originalOwner) else sym`.

retronym committed:
Lambda impl methods static and more stably named

The body of lambdas is compiled into a synthetic method
in the enclosing class. Previously, this method was a public
virtual method named `fully$qualified$Class$$anonfun$n`.
For lambdas that didn't capture a `this` reference, a static
method was used.

This commit changes two aspects.

Firstly, all lambda impl methods are now emitted static.
An extra parameter is added to those that require a this
reference.

This is an improvement as it:

  - allows, shorter, more readable names for the lambda impl method
  - avoids pollution of the vtable of the class. Note that javac uses
    private instance methods, rather than public static methods. If
    we followed its lead, we would be unable to support important use
    cases in our inliner

Secondly, the name of the enclosing method has been included in
the name of the lambda impl method to improve debuggability and
to improve serialization compatibility. The serialization improvement
comes from the way that fresh names for the impl methods are
allocated: adding or removing lambdas in methods not named "foo" won't
change the numbering of the `anonfun$foo$n` impl methods from methods
named "foo". This is in line with user expectations about anonymous
class and lambda serialization stability. Brian Goetz has described
this tricky area well in:

  http://cr.openjdk.java.net/~briangoetz/eg-attachments/lambda-serialization.html

This commit doesn't go as far a Javac, we don't use the hash of the
lambda type info, param names, etc to map to a lambda impl method name.
As such, we are more prone to the type-1 and -2 failures described there.
However, our Scala 2.11.8 has similar characteristics, so we aren't going
backwards.

Special case in the naming: Use "new" rather than "<init>" for constructor enclosed
lambdas, as javac does.

I have also changed the way that "delambdafy target" methods are identifed.
Rather than relying on the naming convention, I have switched to using a
symbol attachment. The assumption is that we only need to identify them
from within the same compilation unit.

This means we can distinguish impl metbods for expanded functions
(ones called from an `apply` method of an ahead-of-time expanded
anonfun class), from those that truly end up as targets for lambda
metafactory. Only the latter are translated to static methods in
this patch.

retronym commented on src/compiler/scala/tools/nsc/ast/TreeGen.scala:
```diff
@@ -310,6 +321,40 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
+    *       - are associating the RHS with a cloned symbol, but intend for the original
+    *         method to remain and for recursive calls to target it.
+    */
+  final def mkStatic(orig: DefDef, maybeClone: Symbol => Symbol): DefDef = {
```
The flexibility of `maybeClone` is to support a forthcoming PR that will reuse this method to emit static methods in traits.

retronym commented:
A closer look at the failure in the optimizer test: https://gist.github.com/a50a921b70944811bdd55268f1942c6b

retronym commented:
Needed to update the regex in `BackendUtils.anonfunAdaptedName`, in addition to the method names in the expected bytecode summaries. Still getting an extra null check.

retronym commented on test/junit/scala/issues/OptimizedBytecodeTest.scala:
```diff
@@ -56,9 +55,9 @@ class OptimizedBytecodeTest extends ClearAfterClass {
-    assertSameSummary(getSingleMethod(c, "C$$$anonfun$1"), List(LDC, "C$$$anonfun$2", IRETURN))
-    assertSameSummary(getSingleMethod(c, "C$$$anonfun$2"), List(-1 /*A*/, GOTO /*A*/))
+      LDC, ASTORE, ALOAD /*0*/, ALOAD /*1*/, "$anonfun$t$1", IRETURN))
+    assertSameSummary(getSingleMethod(c, "$anonfun$t$1"), List(ALOAD, IFNONNULL /*8*/, ACONST_NULL, ATHROW, -1 /*8*/, LDC, "$anonfun$t$2", IRETURN))
```
We could avoid the null check by teaching the backend that the `$this` parameter is non-nullable.

lrytz replied:
here's a wip: https://github.com/lrytz/scala/commit/bb9d0b8f0815026f6c01f7cf7bd4379d906d22c4

retronym commented:
I've had to disable this for lambdas that don't use `invokedynamic`, but instead still need to use an anonymous class. The code that implements needs some more polish / documentation, which I'll add before we merge.

lrytz commented on src/compiler/scala/tools/nsc/ast/TreeGen.scala:
```diff
@@ -269,13 +269,25 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
       val samMethType = fun.tpe memberInfo sam
-      mkMethodForFunctionBody(localTyper)(owner, fun, sam.name.toTermName)(methParamProtos = samMethType.params, resTp = samMethType.resultType)
+      val res = mkMethodForFunctionBody(localTyper)(owner, fun, sam.name.toTermName)(methParamProtos = samMethType.params, resTp = samMethType.resultType)
+      res
```
- [x] debugging artifact?

lrytz replied:
ping here

lrytz commented on src/compiler/scala/tools/nsc/ast/TreeGen.scala:
```diff
@@ -310,6 +322,41 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
+    * Create a new `DefDef` based on `orig` with an explicit self parameter.
+    *
+    * Details:
+    *   - Must by runs after erasure
```
`by runs` -> `be run`?

lrytz commented on src/compiler/scala/tools/nsc/ast/TreeGen.scala:
```diff
@@ -310,6 +322,41 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
+    })
+    val selfParam = ValDef(selfParamSym)
+    val oldSyms = origParams
+    val newSyms = newSym.info.params.drop(1)
```
maybe call it `newParams`

lrytz commented on src/compiler/scala/tools/nsc/ast/TreeGen.scala:
```diff
@@ -310,6 +322,41 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
+      case mt @ MethodType(params, res) => copyMethodType(mt, selfParamSym :: params, res)
+    })
+    val selfParam = ValDef(selfParamSym)
+    val oldSyms = origParams
```
`oldSyms` is just an alias for `origParams`, not needed

lrytz commented on src/compiler/scala/tools/nsc/ast/TreeGen.scala:
```diff
@@ -310,6 +322,41 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
+    *   - Must by runs after erasure
+    *   - If `maybeClone` is the identity function, this runs "in place"
+    *     and mutates the symbol of `orig`. `orig` should be discarded
+    *   - Symbol owners and returns are substituted
```
`returns` -> `parameters`

retronym replied:
Actually `changeOwner` patches up `return`-s, so it should be `owners, returns, and param symbols`

lrytz commented on src/compiler/scala/tools/nsc/ast/TreeGen.scala:
```diff
@@ -310,6 +322,41 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
+    val newSyms = newSym.info.params.drop(1)
+
+    val rhs = orig.rhs.substituteThis(newSym.owner, atPos(newSym.pos)(gen.mkAttributedIdent(selfParamSym)))
+      .substituteSymbols(oldSyms, newSyms).changeOwner(origSym -> newSym)
```
here we are mutating the `rhs` of the original `DefDef` (`substituteSymbols`, `changeOwner`). IUC the the original `DefDef` should remain usable.

lrytz commented on src/compiler/scala/tools/nsc/transform/UnCurry.scala:
```diff
@@ -348,9 +352,13 @@ abstract class UnCurry extends InfoTransform
      */
     private def translateSynchronized(tree: Tree) = tree match {
-      case dd @ DefDef(_, _, _, _, _, Apply(fn, body :: Nil)) if isSelfSynchronized(dd) =>
+      case dd @ DefDef(_, _, _, _, _, Apply(fn, body :: Nil)) if isSelfSynchronized(dd) && dd.symbol.isDelambdafyFunction =>
```
missing a `!` ? and maybe you could add the condition into `isSelfSynchronized`, which already performs some additional checks.

lrytz committed:
Treat self parameter as non-null in the optimizer

lrytz commented on test/junit/scala/tools/nsc/backend/jvm/analysis/ProdConsAnalyzerTest.scala:
```diff
@@ -181,7 +181,7 @@ class ProdConsAnalyzerTest extends ClearAfterClass {
     val List(add)   = findInstr(m, "IADD")
     val List(swap)  = findInstr(m, "SWAP")
-    val List(store) = findInstr(m, "ISTORE")
+    val store = getInstr(m, "ISTORE")
```
there's a few more just above that could be replaced

retronym replied:
Ha, the extra spaces after the `=` fooled my regex :)

lrytz commented on test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala:
```diff
@@ -1501,16 +1499,48 @@ class InlinerTest extends ClearAfterClass {
+  }
+
+  @Test
+  @Ignore // TODO figure out exactly why this fails
```
did it fail on CI or only locally? if you don't go through a full bootstrap (no locker) the failure is expected

lrytz replied:
(STARR / 2.12.0-M4 inlines the wrong method due to https://github.com/scala/scala/pull/5140)

retronym replied:
It failed locally in the SBT build.

lrytz replied:
so it should be fine on the ci

retronym replied:
Right. But shouldn't we eventually restrict ourselves to writing JUnit tests that work under the constraints of the SBT build?

lrytz replied:
yes, it will prevent confusions like in this case.. it will work once we have a new starr though :)

lrytz commented:
the test failure (`tailcalls` for synchronized method) might be due to https://github.com/scala/scala/pull/5157#discussion-diff-63159265

retronym commented:
Rebased to resolve conflicts with @lrytz's recent refactorings to the JUnit test utilities.

lrytz commented on src/compiler/scala/tools/nsc/ast/TreeGen.scala:
```diff
@@ -269,13 +269,25 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
+  def mkLiftedFunctionBodyMethod(localTyper: analyzer.Typer)(owner: Symbol, fun: Function, newNamePrefix: TermName) = {
+    def nonLocalEnclosingMember(sym: Symbol) = {
+      if (sym.isLocalDummy) sym.enclClass.primaryConstructor else {
+        if (sym.isLocalToBlock) sym.originalOwner else sym
```
- [x] i still think we need a recursive call to `nonLocalEnclosingMember` here

lrytz commented on src/compiler/scala/tools/nsc/ast/TreeGen.scala:
```diff
@@ -310,6 +322,38 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
+      case mt @ MethodType(params, res) => copyMethodType(mt, selfParamSym :: params, res)
+    })
+    val selfParam = ValDef(selfParamSym)
+    val rhs = orig.rhs.substituteThis(newSym.owner, atPos(newSym.pos)(gen.mkAttributedIdent(selfParamSym)))
```
all `this` references will be replaced by the same tree (same object reference), as the `tree` parameter of `substituteThis` is not by-name: https://github.com/scala/scala/blob/2.12.x/src/reflect/scala/reflect/internal/Trees.scala#L184-L185

I guess this is an oversight, because the `ThisSubstituter` class does take a by-name param (https://github.com/scala/scala/blob/2.12.x/src/reflect/scala/reflect/internal/Trees.scala#L1541).

retronym replied:
The lack of by-name seems to date back to https://github.com/scala/scala/commit/a03c34c140709e95bb6ab6a8984747b83d630eea#diff-092465703161e9bcdd19dd4d0ca51221R131

Agree it appears to be an oversight.

The tree sharing is probably benign after erasure (we won't clear the type then retype it in one occurence that leave the wrong result for another context). 
- [x] But I'll modify the API in any case.

lrytz commented on src/partest-extras/scala/tools/partest/ASMConverters.scala:
```diff
@@ -2,7 +2,7 @@ package scala.tools.partest
 import scala.collection.JavaConverters._
 import scala.tools.asm
-import asm.{tree => t}
+import asm.{Opcodes, tree => t}
```
no-op

lrytz commented on src/compiler/scala/tools/nsc/transform/UnCurry.scala:
```diff
@@ -213,19 +213,23 @@ abstract class UnCurry extends InfoTransform
         gen.expandFunction(localTyper)(fun, inConstructorFlag)
       } else {
+        val mustExpand = mustExpandFunction(fun)
+        val newNamePrefix = if (mustExpand) nme.DELAMBDAFY_LAMBDA_CLASS_NAME else nme.ANON_FUN_NAME
```
so if we expand to an anon class the method is called `$lambda`, if we use LFM it's named `$anonfun` - why that distinction?

retronym replied:
I need to document this. I needed to exclude non LMF functions/SAMs from the static-rewrite in `Delambdafy`, which runs for methods that satisfy `isDelambdafyTarget`.

Ideally we'd be able to rewrite them as well, but it gets more involved as we need to rewrite the anonymous class that delegates to this method in a non-trivial way.
- [x] Maybe a tree or symbol attachment would be a cleaner way to implement `isDelambdafyTarget`.

lrytz commented:
did you have a chance to look at https://github.com/lrytz/scala/commit/bb9d0b8f0815026f6c01f7cf7bd4379d906d22c4? fixes nullness analysis for the `$this` parameter.

retronym commented:
> did you have a chance to look at lrytz@bb9d0b8? fixes nullness analysis for the $this parameter.

Yep, that looked-GTM. I'll include it here.

retronym committed:
Avoid tree sharing with substituteThis

The underlying transformer has a by-name parameter for the
to provide the `to` tree, but this was strict in the layers
of API above.

Tree sharing is frowned upon in general as it leads to cross
talk when, e.g., the erasure typechecker mutates the
`tpe` field of the shared tree in different context.

retronym commented:
@lrytz Ready for re-review

lrytz commented:
LGTM!

retronym commented:
@lrytz My change to use symbol attachments to identify methods that are delambdafy targets interacted poorly with specialization. I've pushed a slightly modified version with a hack in `Duplicators` to make it work.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Duplicators.scala:
```diff
@@ -229,7 +229,12 @@ abstract class Duplicators extends Analyzer {
+          // bodies in specialized subclasses.
+          if (ddef.hasAttachment[DelambdafyTarget.type])
+            result.symbol.updateAttachment(DelambdafyTarget)
+          result
```
i see the problem.. i don't really understand how duplicators work. first it seems to set `tree.symbol = NoSymbol` and then call `super.typed` - where is the new symbol created and assigned?

adriaanm replied:
I believe that's done in calls to the `retyped` method.

lrytz replied:
i don't really see how that works yet..

adriaanm replied:
It's (probably) a little too clever: the invalidSyms map tracks the old symbol to the new one by mapping to the tree, whose symbol field will be mutated later to the new symbol...

adriaanm replied:
```
  private val invalidSyms: mutable.Map[Symbol, Tree] = perRunCaches.newMap[Symbol, Tree]()
...
   /** Return the new symbol corresponding to `sym`. */
    private def updateSym(sym: Symbol): Symbol =
      if (invalidSyms.isDefinedAt(sym))
        invalidSyms(sym).symbol
      else
        sym
....
          case vdef @ ValDef(mods, name, _, rhs) if mods.hasFlag(Flags.LAZY) =>
            debuglog("ValDef " + name + " sym.info: " + vdef.symbol.info)
            invalidSyms(vdef.symbol) = vdef
            val newowner = owner orElse context.owner
            val newsym = vdef.symbol.cloneSymbol(newowner)
            newsym.setInfo(fixType(vdef.symbol.info))
            vdef.symbol = newsym
...
```

lrytz replied:
ok, so we go through `cloneSymbol` which seems to also clone attachments (https://github.com/scala/scala/blob/25165b7ad9c4aafdddb4e871b195a91014a91b72/src/reflect/scala/reflect/internal/Symbols.scala#L1951) - shouldn't cloned sym already have the `DelambdafyTarget` attachment?

adriaanm replied:
I pasted the `ValDef` case because the `invalidSyms` connection is clearer there, sorry for the confusion. For `DefDef`s the symbol is thrown away and the type checker will create a new one, which explains why the attachment goes walkabout.

retronym replied:
Yep, tree attachments make it through, but I need a symbol attachment to support the one place (in `LambdaLift`) where we need to know if the method we're invoking is a delambdafy target. :sigh:

retronym replied:
... although, I could get rid of this with:

``` diff
diff --git a/src/compiler/scala/tools/nsc/transform/Delambdafy.scala b/src/compiler/scala/tools/nsc/transform/Delambdafy.scala
index 65449b8..c3ba4f4 100644
--- a/src/compiler/scala/tools/nsc/transform/Delambdafy.scala
+++ b/src/compiler/scala/tools/nsc/transform/Delambdafy.scala
@@ -258,7 +258,7 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
         super.transform(transformFunction(fun))
       case Template(_, _, _) =>
         def pretransform(tree: Tree): Tree = tree match {
-          case dd: DefDef if dd.symbol.isDelambdafyTarget =>
+          case dd: DefDef if dd.hasAttachment[DelambdafyTarget.type] =>
             if (!dd.symbol.hasFlag(STATIC) && methodReferencesThis(dd.symbol)) {
               gen.mkStatic(dd, dd.symbol.name, sym => sym)
             } else {
@@ -338,7 +338,7 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
     private var currentMethod: Symbol = NoSymbol

     override def traverse(tree: Tree) = tree match {
-      case DefDef(_, _, _, _, _, _) if tree.symbol.isDelambdafyTarget =>
+      case DefDef(_, _, _, _, _, _) if tree.hasAttachment[DelambdafyTarget.type] =>
         // we don't expect defs within defs. At this phase trees should be very flat
         if (currentMethod.exists) devWarning("Found a def within a def at a phase where defs are expected to be flattened out.")
         currentMethod = tree.symbol
diff --git a/src/compiler/scala/tools/nsc/transform/LambdaLift.scala b/src/compiler/scala/tools/nsc/transform/LambdaLift.scala
index 074acc1..0f40d29 100644
--- a/src/compiler/scala/tools/nsc/transform/LambdaLift.scala
+++ b/src/compiler/scala/tools/nsc/transform/LambdaLift.scala
@@ -523,10 +523,7 @@ abstract class LambdaLift extends InfoTransform {
   } // class LambdaLifter

   private def addFree[A](sym: Symbol, free: List[A], original: List[A]): List[A] = {
-    val prependFree = (
-         !sym.isConstructor // this condition is redundant for now. It will be needed if we remove the second condition in 2.12.x
-      && (settings.Ydelambdafy.value == "method" && sym.isDelambdafyTarget) // SI-8359 Makes the lambda body a viable as the target MethodHandle for a call to LambdaMetafactory
-    )
+    val prependFree = !sym.isConstructor
     if (prependFree) free ::: original
     else             original ::: free
   }
diff --git a/src/compiler/scala/tools/nsc/transform/UnCurry.scala b/src/compiler/scala/tools/nsc/transform/UnCurry.scala
index 374e843..11380b7 100644
--- a/src/compiler/scala/tools/nsc/transform/UnCurry.scala
+++ b/src/compiler/scala/tools/nsc/transform/UnCurry.scala
@@ -222,10 +222,8 @@ abstract class UnCurry extends InfoTransform
           gen.mkForwarder(gen.mkAttributedRef(liftedMethod.symbol), (fun.vparams map (_.symbol)) :: Nil)
         ))

-        if (!mustExpand) {
-          liftedMethod.symbol.updateAttachment(DelambdafyTarget)
+        if (!mustExpand)
           liftedMethod.updateAttachment(DelambdafyTarget)
-        }

         val typedNewFun = localTyper.typedPos(fun.pos)(Block(liftedMethod, super.transform(newFun)))
         if (mustExpand) {
@@ -350,7 +348,7 @@ abstract class UnCurry extends InfoTransform
     private def isSelfSynchronized(ddef: DefDef) = ddef.rhs match {
       case Apply(fn @ TypeApply(Select(sel, _), _), _) =>
         fn.symbol == Object_synchronized && sel.symbol == ddef.symbol.enclClass && !ddef.symbol.enclClass.isTrait &&
-          !ddef.symbol.isDelambdafyTarget /* these become static later, unsuitable for ACC_SYNCHRONIZED */
+          !ddef.hasAttachment[DelambdafyTarget.type] /* these become static later, unsuitable for ACC_SYNCHRONIZED */
       case _ => false
     }

diff --git a/src/reflect/scala/reflect/internal/Symbols.scala b/src/reflect/scala/reflect/internal/Symbols.scala
index 62a33ad..edffe73 100644
--- a/src/reflect/scala/reflect/internal/Symbols.scala
+++ b/src/reflect/scala/reflect/internal/Symbols.scala
@@ -807,7 +807,6 @@ trait Symbols extends api.Symbols { self: SymbolTable =>

     final def isAnonymousFunction = isSynthetic && (name containsName tpnme.ANON_FUN_NAME)
     final def isDelambdafyFunction = isSynthetic && (name containsName tpnme.DELAMBDAFY_LAMBDA_CLASS_NAME)
-    final def isDelambdafyTarget  = isArtifact && isMethod && hasAttachment[DelambdafyTarget.type]
     final def isDefinedInPackage  = effectiveOwner.isPackageClass
     final def needsFlatClasses    = phase.flatClasses && rawowner != NoSymbol && !rawowner.isPackageClass


```

lrytz replied:
> For `DefDefs` the symbol is thrown away and the type checker will create a new one

OK, sorry I'm not giving up yet :) where is that in the typer?

retronym replied:
If we compile:

```
class C[@specialized A] {
  def foo(a: A) = { def bar = 42; a }
}

```

We get to:

https://github.com/scala/scala/blob/92affd3e76d1a3019f1c35ac20af32360d914641/src/compiler/scala/tools/nsc/typechecker/Duplicators.scala#L167-L167

For:

```
tree: def bar(): Int = 42
```

From a call stack:

```
"main@1" prio=5 tid=0x1 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
      at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.invalidate(Duplicators.scala:167)
      at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.scala$tools$nsc$typechecker$Duplicators$BodyDuplicator$$$anonfun$18(Duplicators.scala:181)
      at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.scala$tools$nsc$typechecker$Duplicators$BodyDuplicator$$$anonfun$18$adapted(Duplicators.scala:181)
      at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator$$Lambda$991.478209179.apply(Unknown Source:-1)
      at scala.collection.immutable.List.foreach(List.scala:376)
      at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.invalidateAll(Duplicators.scala:181)
      at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.typed(Duplicators.scala:220)
      at scala.tools.nsc.typechecker.Typers$Typer.transformedOrTyped(Typers.scala:5602)
      at scala.tools.nsc.typechecker.Typers$Typer.typedDefDef(Typers.scala:2243)
      at scala.tools.nsc.typechecker.Typers$Typer.typedMemberDef$1(Typers.scala:5305)
      at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5356)
      at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5393)
      at scala.tools.nsc.typechecker.Typers$Typer.typedInternal(Typers.scala:5423)
      at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5367)
      at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5371)
      at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.typed(Duplicators.scala:232)
      at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5448)
      at scala.tools.nsc.typechecker.Duplicators.retyped(Duplicators.scala:36)
      at scala.tools.nsc.transform.SpecializeTypes$SpecializationDuplicator.super$retyped(SpecializeTypes.scala:1331)
      at scala.tools.nsc.transform.SpecializeTypes$SpecializationDuplicator.scala$tools$nsc$transform$SpecializeTypes$SpecializationDuplicator$$$anonfun$154(SpecializeTypes.scala:1331)
      at scala.tools.nsc.transform.SpecializeTypes$SpecializationDuplicator$$Lambda$963.1043639521.apply(Unknown Source:-1)
      at scala.reflect.internal.SymbolTable.enteringPhase(SymbolTable.scala:235)
      at scala.tools.nsc.Global.enteringSpecialize(Global.scala:967)
      at scala.tools.nsc.transform.SpecializeTypes$SpecializationDuplicator.retyped(SpecializeTypes.scala:1331)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.duplicateBody(SpecializeTypes.scala:1791)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$196(SpecializeTypes.scala:1641)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$950.1527007086.apply(Unknown Source:-1)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.reportError(SpecializeTypes.scala:1445)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transformDefDef$1(SpecializeTypes.scala:1641)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform1(SpecializeTypes.scala:1713)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$164(SpecializeTypes.scala:1453)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$816.778720569.apply(Unknown Source:-1)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.reportError(SpecializeTypes.scala:1445)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1453)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1396)
      at scala.reflect.api.Trees$Transformer.scala$reflect$api$Trees$Transformer$$$anonfun$6(Trees.scala:2559)
      at scala.reflect.api.Trees$Transformer$$Lambda$703.2052435819.apply(Unknown Source:-1)
      at scala.collection.immutable.List.loop$1(List.scala:171)
      at scala.collection.immutable.List.mapConserve(List.scala:187)
      at scala.reflect.api.Trees$Transformer.transformTrees(Trees.scala:2559)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$190(SpecializeTypes.scala:1624)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$911.1692073709.apply(Unknown Source:-1)
      at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
      at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
      at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transformTemplate$1(SpecializeTypes.scala:1624)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform1(SpecializeTypes.scala:1626)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$164(SpecializeTypes.scala:1453)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$816.778720569.apply(Unknown Source:-1)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.reportError(SpecializeTypes.scala:1445)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1453)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1396)
      at scala.reflect.api.Trees$Transformer.transformTemplate(Trees.scala:2563)
      at scala.reflect.internal.Trees.scala$reflect$internal$Trees$$$anonfun$37(Trees.scala:1412)
      at scala.reflect.internal.Trees$$Lambda$699.2089589253.apply(Unknown Source:-1)
      at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
      at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
      at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
      at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:17)
      at scala.reflect.internal.Trees.itransform(Trees.scala:1411)
      at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
      at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
      at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
      at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:44)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform1(SpecializeTypes.scala:1737)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$164(SpecializeTypes.scala:1453)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$816.778720569.apply(Unknown Source:-1)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.reportError(SpecializeTypes.scala:1445)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1453)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1396)
      at scala.reflect.api.Trees$Transformer.scala$reflect$api$Trees$Transformer$$$anonfun$13(Trees.scala:2589)
      at scala.reflect.api.Trees$Transformer$$Lambda$698.1611370719.apply(Unknown Source:-1)
      at scala.collection.immutable.List.loop$1(List.scala:171)
      at scala.collection.immutable.List.mapConserve(List.scala:187)
      at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2587)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$186(SpecializeTypes.scala:1610)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$818.41765385.apply(Unknown Source:-1)
      at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
      at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform1(SpecializeTypes.scala:1608)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$164(SpecializeTypes.scala:1453)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$816.778720569.apply(Unknown Source:-1)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.reportError(SpecializeTypes.scala:1445)
      at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1453)
      at scala.tools.nsc.transform.SpecializeTypes$SpecializationTransformer.scala$tools$nsc$transform$SpecializeTypes$SpecializationTransformer$$$anonfun$269(SpecializeTypes.scala:1983)
      at scala.tools.nsc.transform.SpecializeTypes$SpecializationTransformer$$Lambda$815.1415630650.apply(Unknown Source:-1)
      at scala.reflect.internal.SymbolTable.enteringPhase(SymbolTable.scala:235)
      at scala.reflect.internal.SymbolTable.exitingPhase(SymbolTable.scala:256)
      at scala.tools.nsc.Global.exitingSpecialize(Global.scala:957)
      at scala.tools.nsc.transform.SpecializeTypes$SpecializationTransformer.transform(SpecializeTypes.scala:1983)
      at scala.tools.nsc.transform.SpecializeTypes$SpecializationTransformer.transform(SpecializeTypes.scala:1979)
      at scala.tools.nsc.ast.Trees$Transformer.transformUnit(Trees.scala:140)
      at scala.tools.nsc.transform.Transform$Phase.apply(Transform.scala:30)
      at scala.tools.nsc.Global$GlobalPhase.scala$tools$nsc$Global$GlobalPhase$$$anonfun$11(Global.scala:402)
      at scala.tools.nsc.Global$GlobalPhase$$Lambda$387.888655833.apply$mcV$sp(Unknown Source:-1)
      at scala.tools.nsc.Global$GlobalPhase.withCurrentUnitNoLog(Global.scala:395)
      at scala.tools.nsc.Global$GlobalPhase.withCurrentUnit(Global.scala:386)
      at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:402)
      at scala.tools.nsc.Global$GlobalPhase.scala$tools$nsc$Global$GlobalPhase$$$anonfun$10(Global.scala:366)
      at scala.tools.nsc.Global$GlobalPhase.scala$tools$nsc$Global$GlobalPhase$$$anonfun$10$adapted(Global.scala:366)
      at scala.tools.nsc.Global$GlobalPhase$$Lambda$386.980138431.apply(Unknown Source:-1)
      at scala.collection.Iterator.foreach(Iterator.scala:932)
      at scala.collection.AbstractIterator.foreach(Iterator.scala:1409)
      at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:366)
      at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1396)
      at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1381)
      at scala.tools.nsc.Global$Run.compileSources(Global.scala:1376)
      at scala.tools.nsc.Global$Run.compile(Global.scala:1470)
      at scala.tools.nsc.Driver.doCompile(Driver.scala:35)
      at scala.tools.nsc.MainClass.doCompile(Main.scala:24)
      at scala.tools.nsc.Driver.process(Driver.scala:55)
      at scala.tools.nsc.Driver.main(Driver.scala:68)
      at scala.tools.nsc.Main.main(Main.scala:-1)

```

Note that the delambdafy target method that is still method-local at the specialize phase.

adriaanm replied:
in `invalidate`:

```
          case DefDef(_, name, tparams, vparamss, _, rhs) =>
            // invalidate parameters
            invalidateAll(tparams ::: vparamss.flatten)
            tree.symbol = NoSymbol // <-- so long, symbol!
```

lrytz replied:
i see where the symbol is removed from the tree, but still wonder where the new one is created. anyway, i'm waisting everyone's time - i think i'd prefer the tree-based approach you mention above (https://github.com/scala/scala/pull/5157#discussion_r65637086), but i don't want to hold things back unnecessarily.

retronym replied:
```
    def typedBlock(block0: Block, mode: Mode, pt: Type): Block = {
      val syntheticPrivates = new ListBuffer[Symbol]
      try {
        namer.enterSyms(block0.stats)
```

lrytz replied:
thanks, got it now! i didn't connect the dots that those were local methods.

adriaanm replied:
No problem! It's good to dig deep. I think we'd also create new symbols for non-local methods in `typedTemplate`:

```
enterSyms(context.outer.make(templ, clazz, clazz.info.decls), body1)
```

(it eventually gets to `enterDefDef` in Namers)

retronym committed:
Drop local suffix in lambda impl method name

retronym commented:
I think this one and #5099 are ready to roll. @adriaanm WDYT?

adriaanm commented:
:shipit:!
